File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        cache={}#this will store our key:value(list) pairs
        for word in strs:
            temp=''.join(sorted(word))#temp is the word formed by sorting the letters of the word
            if temp not in cache:
                cache[temp]=[word]#if its the first add to dictionary
            else:
                cache[temp].append(word)#append if already exists
        return list(cache.values())#just return the values

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) return new ArrayList();
        Map<String, List<String>> cache=new HashMap<>();
        List<List<String>> result=new ArrayList<>();
        char[] temp;
        String key;
        for(int i=0;i<strs.length;i++){
            temp=strs[i].toCharArray();
            Arrays.sort(temp);
            key=String.valueOf(temp);
            if(!cache.containsKey(key)){
                cache.put(key, new ArrayList<>(Arrays.asList(strs[i])));
                /*OR
                List<String> stringList= new ArrayList<>();
                stringList.add(strs[i]);
                cache.put(key,stringList);*/
                          
            }else{
                cache.get(key).add(strs[i]);
            }
        }
        
        result.addAll(cache.values());
        return result;
        
    }
}